Record dependencies on macro arguments

Duhemm opened the pull request:
> Macros take arguments as trees and return some other trees; both of them have dependencies but we see trees only after expansion and recorded only those dependencies.

(from [this discussion in sbt-dev](https://groups.google.com/d/msg/sbt-dev/zDOIFDSNRrk/_YTVn7bRJb0J)).

This pull request solves this problem by looking into the attachments of the trees that are supposed to contain originals of macro expansions and recording dependencies of the macro before its expansion.

gkossakowski committed:
Test for missing dependencies of macro arguments

Add a test which shows the problem of not properly capturing dependencies
of macro arguments.

Commit:
Record dependencies on macro arguments

Macros take arguments as trees and return some other trees; both of
them have dependencies but we see trees only after expansion and
recorded only those dependencies.

This commit solves this problem by looking into the attachments of the
trees that are supposed to contain originals of macro expansions and
recording dependencies of the macro before its expansion.

jsuereth commented on sbt/src/sbt-test/source-dependencies/macro-arg-dep/project/build.scala:
```diff
@@ -0,0 +1,29 @@
+	   base = file("."),
+	   id = "macro",
+	   aggregate = Seq(macroProvider, macroClient),
+	   settings = Defaults.defaultSettings ++ defaultSettings
```
Note:  Usage of Defaults.defaultSettings is going to be deprecated in 0.13.2

jsuereth commented:
So this looks extra super cool!  I'm going to have to delay this until 0.13.3 though. I have way too many issues to resolve for 0.13.2 right now (with auto-plugins), but I'll get back to you next week after the 0.13.2-RC1 is out and 0.13 is ok to merge into again.

Duhemm commented:
(ping @xeno-by)

gkossakowski commented on compile/interface/src/main/scala/xsbt/Compat.scala:
```diff
@@ -91,4 +91,39 @@ abstract class Compat
+			import Compat._
+
+			locally {
+				import global._ // this is where MEA lives in 2.10.x
```
Can we import just MEA from Global? Global has more than 1600 members, we don't need to pull all of them here.

Duhemm replied:
Unfortunately this won't compile with Scala 2.8 and 2.9 since `MacroExpansionAttachment` appeared in 2.10.

gkossakowski replied:
Ah, yes. This is the way to go then. Could you just expand the comment in the code explaining why do we need a wildcard import?

gkossakowski commented on compile/interface/src/main/scala/xsbt/Compat.scala:
```diff
@@ -91,4 +91,39 @@ abstract class Compat
+				}
+
+				locally {
+					import analyzer._ // this is where MEA lives in 2.11.x
```
Same as above, let's just import MEA, not the whole Analyzer.

xeno-by replied:
Just to summarize previous discussion. `MEA` can't be imported here, because that would be incompatible with versions prior to 2.11.

gkossakowski commented on compile/interface/src/main/scala/xsbt/Compat.scala:
```diff
@@ -91,4 +91,39 @@ abstract class Compat
+			locally {
+				import global._ // this is where MEA lives in 2.10.x
+				implicit def withExpandee(att: MacroExpansionAttachment): WithExpandee = new WithExpandee(att)
+				class WithExpandee(att: MacroExpansionAttachment) {
```
Why do we need that implicit class? Is it because expandee is defined in 2.11, or what?

Duhemm replied:
Exactly, it has been renamed from `original` to `expandee` in 2.11.

gkossakowski replied:
Ok. Let's document it in the code.

xeno-by commented:
Great job, @Duhemm!

gkossakowski commented:
@jsuereth: don't worry about this PR, I'll take care of it and schedule for sbt 0.13.3 release.

gkossakowski commented:
Indeed. This looks very promising! I'm glad you managed to figure out the maze of compatibility hacks we have here. :thumbsup:

Have you considered expanding the `macro-arg-dep` test to cover nested macro invocations? I have something like that in mind:

``` scala
package macro

object ClientNested {
    Provider.printTree(Provider.printTree(Foo.str))
}
```

Also, given the fact that you modify both `Dependency` and `ExtractUsedNames`, we should expand their units tests:

https://github.com/sbt/sbt/blob/0.13/compile/interface/src/test/scala/xsbt/DependencySpecification.scala
https://github.com/sbt/sbt/blob/0.13/compile/interface/src/test/scala/xsbt/ExtractUsedNamesSpecification.scala

The code samples you are going to use in those tests should look similar to what you have in the `macro-arg-dep`.

Just in case you it's not clear what's the point of testing (seemingly) the same thing twice: scripted tests are [functional tests](http://en.wikipedia.org/wiki/Functional_testing) and the specifications for `Dependency` and `ExtractUsedNames` are [unit tests](http://en.wikipedia.org/wiki/Unit_testing). Here's a concise description of the difference: http://www.softwaretestingtricks.com/2007/01/unit-testing-versus-functional-tests.html

Duhemm commented:
Hi guys,

Does anybody know what happened with the last build ? It seems to have timed out, but I don't see how the changes between 5f6b06a and d76b4ea may have caused this ?

I've had to modify `ScalaCompilerForUnitTesting` so that one can compile separately multiple files (this was required in order to compile macro expansions). What do you think ? I will look into the unit tests for `ExtractUsedNames` shortly !

jsuereth commented:
Yeah, it's transient to you, see the error message:

```
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
```

It's quite possible there's a rare deadlock in the test listener, which would have caused this.  It's the first time I've seen this message, but something I can pay attention to.

xeno-by commented on compile/interface/src/main/scala/xsbt/ExtractUsedNames.scala:
```diff
@@ -53,30 +53,44 @@ class ExtractUsedNames[GlobalType <: CallbackGlobal](val global: GlobalType) {
-			case _ => ()
+
+		def handleTreeNode(node: Tree): Unit = {
+			def handleMacroExpansion(original: Tree): Unit = original.foreach(handleTreeNode)
```
Can we have a test of a nested macro expansion, i.e. a macro application that expands into another macro application?

gkossakowski commented on sbt/src/sbt-test/source-dependencies/macro-arg-dep/test:
```diff
@@ -0,0 +1,15 @@
+
+# we should recompile Foo.scala first and then fail to compile Client.scala due to missing
+# `Foo.str`; however recompilation of Client.scala is never triggered due to missing
+# dependency
```
We fix this problem in this PR so let's update the comment to say that it does trigger recompilation because dependencies are properly recorded.

gkossakowski commented:
Ok. I think we just missing test for nested macro applications we'll be done with the code. Once the last patch (for nested macro applications) is submitted we should look into squashing most of the commits. E.g. 14ff4b0 and c9be404 squashed reduce to empty commit as the latter one cancels the former. All small commits that fix typos, reduce use of reflection, etc. should be squashed as well.

Commit:
Add scripted test for nested macros

Add test analogous to source-dependencies/macro-arg-dep
but check if dependencies of nested macro applications are handled
properly. Nested macro applications are tricky because we have to look
into original (before macro expansion) trees recursively. This test verifies
that.

Duhemm commented:
Hi,

I just added this test and updated the comments you mentioned, is everything ok ?

gkossakowski commented:
Yes, looks good. Could you rebase the branch and squash most of the commits? I think we would like to keep commits for the following events:
- add pending test as in 2bd07bd
- record depedencies on macro arguments as in 8d1e409
- extend unit testing infrastructure in separate commit and then use it, it would mean that you break up 5f6b06a into two (this is optional step if you know how to split commits in Git)
- add tests for nested macros

That would give us 5 commits in total.

Commit:
Improve unit testing compiler

It was not possible to make `ScalaCompilerForUnitTesting` compile several
files in different runs, which means that it was not possible to compile
and use a macro in a test case, since macros cannot be used in the same
compilation run that defines them.

This commit allows a test case to provide multiple grouped snippets of
code that will be compiled in separate runs.

For instance :
    List(Map(<snippet A>, <snippet B>), Map(<snippet C>))

Here, <snippet A> and <snippet B> will be compiled together, and then
<snippet C> will be compiled, and will be able to use symbols defined
in <snippet A> or <snippet B>.

Commit:
Unit test for dependency extraction from macro applications

Add a unit test which checks whether we capture dependencies introduced
by arguments to macros. Those dependencies are special because macros
get expanded during type checking and arguments to macros are not visible
during regular tree walk.

Duhemm commented:
It looks like I managed to split the commits ! Thank you for your help !

gkossakowski commented:
@jsuereth: since this PR is shaping very nicely I'd like to get it in sbt 0.13.2. It's a very nice piece of work and improves significantly macro handling in name hashing. It's safe because it affects name hashing only. The old incremental compilation algorithm is unaffected. WDYT?

gkossakowski commented:
@Duhemm: you need to update one more commit :)

gkossakowski commented:
LGTM

gkossakowski commented:
@jsuereth: ping on merging this into 0.13.2

xeno-by commented:
\o/

eed3si9n commented:
I'm merging this into 0.13.2. @gkossakowski Could you add an entry into https://github.com/sbt/sbt/blob/0.13/src/sphinx/Community/Changes.rst to summarize this enhancement, plz?
